### PR TITLE
Allow loading of images via OpenCV

### PR DESCRIPTION
**Pull request by seth-planet**
Image resizing is so much faster with OpenCV due to threading and vector optimizations. In theory it should be as fast as 'pillow-simd', but in practice seems to be even 4x faster than 'pillow-simd'. On my Mac, OpenCV resizes 10x faster than default pillow (non-SIMD). Unfortunately, the OpenCV package is also 10x the size of pillow, too.

Note: PR currently proposed and untested.

For performance testing, see my post here:
https://ipcamtalk.com/threads/codeproject-ai-version-2-5.74304/page-21#post-846822

### CODE REVIEW

File: src/SDK/Python/request_data.py [hunk 1]
```diff
@@ -150,16 +159,19 @@ def get_image(self, index : int) -> Image:
...
-            img_file    = self.files[index]
-            img_dataB64 = img_file["data"]
-            img_bytes   = base64.b64decode(img_dataB64)
-            
-            with io.BytesIO(img_bytes) as img_stream:
-                img = Image.open(img_stream).convert("RGB")
-                return img
+            if module == 'opencv':
+                assert 'cv' in sys.modules and 'np' in sys.modules
+                return cv.imdecode(np.frombuffer(img_stream, dtype=np.uint8), cv.IMREAD_COLOR)
```
Comment (ChrisMaunder) [inline src/SDK/Python/request_data.py:168]: img_stream hasn't yet been declared or assigned.

File: src/SDK/Python/request_data.py [hunk 2]
```diff
@@ -135,11 +143,12 @@ def add_file(self, file_name: str) -> None:
             return
         self.payload["files"].append({ "data": RequestData.encode_file_contents(file_name) })
 
-    def get_image(self, index : int) -> Image:
+    def get_image(self, index : int, module: str = 'pil'):
```
Comment (ChrisMaunder) [inline src/SDK/Python/request_data.py:146]: I would add `-> Union[Image, numpy.ndarray]:` as a return type hint

**Review by ChrisMaunder — CHANGES_REQUESTED** [review]
I love this idea: Just a couple of reviews though. The code doesn't seem to work as-is.

**Comment by seth-planet** [comment]
Good to hear you’re on board with it; should be fixed now! ~Sethvia iPhoneOn Apr 26, 2024, at 5:57 AM, ChrisMaunder ***@***.***> wrote:﻿
@ChrisMaunder requested changes on this pull request.

I love this idea: Just a couple of reviews though. The code doesn't seem to work as-is.

In src/SDK/Python/request_data.py:
>                  return None
 
-            img_file    = self.files[index]
-            img_dataB64 = img_file["data"]
-            img_bytes   = base64.b64decode(img_dataB64)
-            
-            with io.BytesIO(img_bytes) as img_stream:
-                img = Image.open(img_stream).convert("RGB")
-                return img
+            if module == 'opencv':
+                assert 'cv' in sys.modules and 'np' in sys.modules
+                return cv.imdecode(np.frombuffer(img_stream, dtype=np.uint8), cv.IMREAD_COLOR)

img_stream hasn't yet been declared or assigned.



In src/SDK/Python/request_data.py:
> @@ -135,11 +143,12 @@ def add_file(self, file_name: str) -> None:
             return
         self.payload["files"].append({ "data": RequestData.encode_file_contents(file_name) })
 
-    def get_image(self, index : int) -> Image:
+    def get_image(self, index : int, module: str = 'pil'):

I would add -> Union[Image, numpy.ndarray]: as a return type hint

**Comment by ChrisMaunder** [comment]
Just merging this and will try and test tonight. Too much on my plate today. Getting worried beer o’clock could be compromised today.

cheers
Chris

**Comment by seth-planet** [comment]
We should schedule a zoom hangout around beer o'clock sometime!

**Comment by MikeLud** [comment]
> We should schedule a zoom hangout around beer o'clock sometime!

Count me in.

**Comment by ChrisMaunder** [comment]
I've merged the changes outside this pull request due to the faffing around we were doing with splitting out the repos.  Will organise zoom call for this coming week

**Pull request closed without merging**